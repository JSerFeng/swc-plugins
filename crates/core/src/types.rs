use std::collections::HashMap;

use plugin_modularize_imports::PackageConfig;
use plugin_import::PluginImportConfig;
use plugin_react_utils::ReactUtilsConfig;
use shared::swc::config::Options;

#[derive(Default)]
pub struct Extensions {
  pub modularize_imports: Option<HashMap<String, PackageConfig>>,
  pub plugin_import: Option<Vec<PluginImportConfig>>,
  pub react_utils: Option<ReactUtilsConfig>,
}

pub struct TransformConfig {
  pub swc: Options,

  /// Internal rust-swc Plugins
  pub extensions: Extensions,
}
