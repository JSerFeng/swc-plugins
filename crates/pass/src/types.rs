use std::collections::HashMap;

use modularize_imports::PackageConfig;
use plugin_import::{PluginImportConfigNapi, PluginImportConfig};
use shared::{napi, napi_derive::napi, swc::config::Options};

/**
 * Internal plugin
 */
#[derive(Default)]
#[napi(object)]
pub struct ExtensionsNapi {
  pub modularize_imports: Option<HashMap<String, PackageConfig>>,
  pub plugin_import: Option<Vec<PluginImportConfigNapi>>,
}

#[derive(Default)]
pub struct Extensions {
  pub modularize_imports: Option<HashMap<String, PackageConfig>>,
  pub plugin_import: Option<Vec<PluginImportConfig>>,
}

#[napi(object)]
pub struct TransformConfigNapi {
  /// Raw swc options
  #[napi(ts_type = "import('./types').SwcOptions")]
  pub swc: String,

  /// Internal rust-swc Plugins
  pub extensions: ExtensionsNapi,
}

pub struct TransformConfig {
  pub swc: Options,

  /// Internal rust-swc Plugins
  pub extensions: Extensions,
}
