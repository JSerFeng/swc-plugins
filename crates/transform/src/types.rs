use shared::{
  serde::Deserialize,
  swc::config::{Options},
};

#[derive(Debug, Deserialize)]
pub struct Extensions {
    pub plugin_import: Option<Vec<PluginImportItem>>
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TransformConfig {
  /// Raw swc options
  pub swc: Options,

  /// Internal rust-swc Plugins
  pub extensions: Extensions
}

#[derive(Debug, Deserialize)]
pub struct PluginImportItem {
  pub source: String,

  pub transform_es: String,
  pub transform_style: String,

  pub ignore_components: Option<Vec<String>>,
  pub lower: bool,
}
