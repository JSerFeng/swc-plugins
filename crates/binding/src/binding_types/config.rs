use core::types::TransformConfig;

use napi::bindgen_prelude::Buffer;
use napi_derive::napi;
use shared::serde_json;

use super::{extensions::ExtensionsNapi, FromNapi};

#[napi(object)]
pub struct TransformConfigNapi {
  /// Raw swc options
  pub swc: Buffer,

  /// Internal rust-swc Plugins
  pub extensions: ExtensionsNapi,
}

impl FromNapi<TransformConfig> for TransformConfigNapi {
  fn from_napi(self, env: napi::Env) -> napi::Result<TransformConfig> {
    let Self { swc, extensions } = self;

    Ok(TransformConfig {
      swc: serde_json::from_slice(&swc.to_vec().as_slice()).unwrap(),
      extensions: extensions.from_napi(env)?,
    })
  }
}
