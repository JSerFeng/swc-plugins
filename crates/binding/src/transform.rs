use napi::{bindgen_prelude::AsyncTask, JsObject, Status, Task};
use shared::{napi, napi_derive::napi, serde_json, swc::TransformOutput};
use transform::types::TransformConfig;

pub struct Transformer {
  config: String,
  code: String,
}

impl Task for Transformer {
  type Output = TransformOutput;
  type JsValue = JsObject;

  fn compute(&mut self) -> napi::Result<Self::Output> {
    let config: TransformConfig = serde_json::from_str(&self.config).unwrap();

    core::transform::transform(config, &self.code)
      .map_err(|e| napi::Error::new(Status::GenericFailure, e.to_string()))
  }

  fn resolve(&mut self, env: napi::Env, output: Self::Output) -> napi::Result<Self::JsValue> {
    let mut obj: JsObject = env.create_object()?;
    obj.set_named_property("code", env.create_string(&output.code)?)?;
    if let Some(map) = output.map {
      obj.set_named_property("map", env.create_string(&map)?)?;
    }
    Ok(obj)
  }
}

// type def here doesn't matter that much, because @modern-js/swc would override this
#[napi(ts_return_type = "Promise<{ code: string, map?: string }>")]
pub fn transform(config: String, code: String) -> AsyncTask<Transformer> {
  AsyncTask::new(Transformer { config, code })
}
