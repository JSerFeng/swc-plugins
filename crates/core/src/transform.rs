use crate::{get_compiler, pass::internal_transform_pass};
use shared::{
  anyhow::Result,
  swc::{self, try_with_handler, HandlerOpts, TransformOutput},
  swc_common::{FileName, Mark},
  swc_ecma_transforms_base::pass::noop,
  swc_ecma_visit::as_folder,
};

use transform::types::TransformConfig;

pub fn transform(config: TransformConfig, code: &str) -> Result<TransformOutput> {
  let compiler = get_compiler();
  let cm = compiler.cm.clone();

  try_with_handler(cm.clone(), HandlerOpts::default(), |handler| {
    compiler.run_transform(handler, true, || {
      let source_filename = "test";
      let fm = cm.new_source_file(
        FileName::Custom(source_filename.to_string()),
        code.to_string(),
      );

      // main transform pass
      let top_level_mark = Mark::new();

      compiler.process_js_with_custom_pass(
        fm,
        None,
        handler,
        &swc::config::Options {
          top_level_mark: Some(top_level_mark),
          ..config.swc.clone()
        },
        |_, _| as_folder(internal_transform_pass(cm, &config, top_level_mark)),
        |_, _| noop(),
      )
    })
  })
}
