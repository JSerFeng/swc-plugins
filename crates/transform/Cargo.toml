[package]
name = "transform"
version = "0.1.0"
edition = "2021"

[dependencies]
regex = "1.6.0"
shared = { path = "../shared" }
serde = "1.0.143"
