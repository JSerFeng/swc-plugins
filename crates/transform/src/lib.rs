pub mod types;
pub mod babel_import;
pub mod noop_pass;
pub mod parser;
pub mod visit;
pub mod react;