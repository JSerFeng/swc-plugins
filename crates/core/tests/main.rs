use core::transform::transform;

use pass::types::Extensions;
use shared::swc::config::Options;

#[test]
fn test() {
  let code = "const a = {};
  const b = { ...a }";
  let mut extension: Extensions = Default::default();

  let config = pass::types::TransformConfig {
    swc: Options {
      config: shared::serde_json::from_str(
        r#"{
        "jsc": {
          "externalHelpers": true
        }
      }"#,
      )
      .unwrap(),
      ..Default::default()
    },
    extensions: &mut extension,
  };

  let res = transform(None, config, code).unwrap();
  insta::assert_snapshot!("plugin-import", res.code);
}
