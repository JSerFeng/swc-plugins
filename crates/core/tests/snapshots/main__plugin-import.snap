---
source: crates/core/tests/main.rs
expression: res.code
---
import _object_spread from "@swc/helpers/src/_object_spread.mjs";
var a = {};
var b = _object_spread({}, a);

